addons/hr_holidays/models/res_company.py

```diff
@@ -0,0 +1,135 @@
+                })
+                continue
+
+            company_subdiv = company.state_id.code
+            subdiv = subdivision_per_country_dict.get(company.country_code, None)
+            user_lang_iso_code = self.env["res.lang"]._lang_get(self.env.user.lang).iso_code[:2]
+            if not user_lang_iso_code or user_lang_iso_code not in lang_per_country.get(company.country_code, []):
+                user_lang_iso_code = 'en_US'
+
+            public_holiday_dict = holidays.country_holidays(
+                company.country_code,
+                subdiv=company_subdiv if subdiv and company_subdiv in subdiv else None,
+                years=year_range,
+                language=user_lang_iso_code,
+            )
+
+            overlapped_holidays = False
+            company_tz = pytz.timezone(company.resource_calendar_id.tz)
+            public_holidays_values_dict = {}
+
+            for holiday_date, holiday_name in public_holiday_dict.items():
+                holiday_start_utc, holiday_end_utc = self._convert_to_utc(holiday_date, holiday_date, company_tz)
+                overlapping = any(
+                    date_from <= holiday_end_utc and
+                    date_to >= holiday_start_utc
+                    for date_from, date_to in existing_holidays_dict.get(company.id, [])
+                )
+                if overlapping:
+                    overlapped_holidays = True
+                    continue
+                if holiday_date in public_holidays_values_dict:
+                    public_holidays_values_dict[holiday_date]['name'] += f" / {holiday_name}"
+                else:
+                    public_holidays_values_dict[holiday_date] = {
+                        'name': holiday_name,
+                        'date_from': holiday_start_utc,
+                        'date_to': holiday_end_utc,
+                        'company_id': company.id,
+                    }
+
+            self.env['resource.calendar.leaves'].with_context(convert_datetime=convert_datetime).create(
+                list(public_holidays_values_dict.values()),
+            )
+            if overlapped_holidays:
+                response.append({
+                    'title': self.env._('Public Holidays Import Notification'),
+                    'type': 'success',
+                    'message': self.env._('Some Public holidays for the next %(years)s years already exist — Some partial changes could have been done.', years=len(year_range)),
+                })
+            else:
+                response.append({
+                    'title': self.env._('Public Holidays Loaded'),
+                    'type': 'success',
+                    'message': self.env._('Public holidays have been successfully created for %(company)s for the next %(years)s years.', company=company.name, years=len(year_range)),
+                })
```

Comment:
That way we get even more cases coverage
```suggestion
            new_pto = self.env['resource.calendar.leaves'].with_context(convert_datetime=convert_datetime).create(
                list(public_holidays_values_dict.values()),
            )
            notification = {
                'title': self.env._('Public Holidays Import Notification'),
                'type': 'success',
                'message': self.env._('No new public time off were added as they already exist.'),
            })
            if not new_pto:
                response.append(notification)
            else:
                 notification['message'] = self.env._(
                        "Public holidays have been successfully created for %(company)s for the next %(years)s years.",
                        company=company.name, years=len(year_range))
                if overlapped_holidays:
                    notification['message'] += " " + self.env._("Some were overlapping existing ones, not all records have been created.")
                response.append(notification)
```